homeassistant/components/sensor/nissan_leaf.py

```diff
@@ -0,0 +1,116 @@
+"""Battery Charge and Range Support for the Nissan Leaf.
+
+Please refer to the main platform component for configuration details
+"""
+
+import logging
+from homeassistant.util.unit_system import IMPERIAL_SYSTEM, METRIC_SYSTEM
+from homeassistant.util.distance import LENGTH_KILOMETERS, LENGTH_MILES
+from homeassistant.helpers.icon import icon_for_battery_level
+from homeassistant.components.nissan_leaf import (
+    DATA_BATTERY, DATA_CHARGING, DATA_LEAF, DATA_RANGE_AC, DATA_RANGE_AC_OFF,
+    LeafEntity
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['nissan_leaf']
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Sensors setup."""
+    _LOGGER.debug("setup_platform nissan_leaf sensors, discovery_info=%s",
+                  discovery_info)
+
+    devices = []
+    for key, value in hass.data[DATA_LEAF].items():
+        _LOGGER.debug("adding sensor for item key=%s, value=%s", key, value)
+        devices.append(LeafBatterySensor(value))
+        devices.append(LeafRangeSensor(value, True))
+        devices.append(LeafRangeSensor(value, False))
+
+    add_devices(devices, True)
+
+
+class LeafBatterySensor(LeafEntity):
+    """Nissan Leaf Battery Sensor."""
+
+    @property
+    def name(self):
+        """Sensor Name."""
+        return self.car.leaf.nickname + " Charge"
+
+    @property
+    def device_class(self):
+        """Return the device class of the sensor."""
+        return 'battery'
+
+    @property
+    def state(self):
+        """Battery state percentage."""
+        return round(self.car.data[DATA_BATTERY], 0)
+
+    @property
+    def unit_of_measurement(self):
+        """Battery state measured in percentage."""
+        return '%'
+
+    @property
+    def icon(self):
+        """Battery state icon handling."""
+        chargestate = self.car.data[DATA_CHARGING]
+        return icon_for_battery_level(
+            battery_level=self.state,
+            charging=chargestate
+        )
+
+
+class LeafRangeSensor(LeafEntity):
+    """Nissan Leaf Range Sensor."""
+
+    def __init__(self, car, ac_on):
+        """Set-up range sensor. Store if AC on."""
+        self.ac_on = ac_on
+        super().__init__(car)
+
+    @property
+    def name(self):
+        """Update sensor name depending on AC."""
+        if self.ac_on is True:
```

Comment:
Name of an entity can't be dynamic! It breaks any automation

Reply:
`ac_on` is only set during the constructor of a LeafRangeSensor.  There's two sensors, one for the range with air conditioning off, and one for air conditioning on. Both or which are always instantiated. I will change `self.ac_on` to `self._ac_on` to indicate it's an internal property.

Reply:
Change made to `_ac_on` in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/3c7c509506dfa4e39afed8eecfbd468569fba004